Tribler/Core/TorrentChecker/session.py

```diff
@@ -198,309 +185,384 @@ def __init__(self, tracker_url, tracker_address, announce_page, on_result_callba
             for infohash in response_dict['files']:
-                downloaded = response_dict['files'][infohash].get('downloaded', 0)
                 complete = response_dict['files'][infohash].get('complete', 0)
+                downloaded = response_dict['files'][infohash].get('downloaded', 0)
                 incomplete = response_dict['files'][infohash].get('incomplete', 0)
 
-                seeders = downloaded
+                # Sow complete as seeders. "complete: number of peers with the entire file, i.e. seeders (integer)"
+                #  - https://wiki.theory.org/BitTorrentSpecification#Tracker_.27scrape.27_Convention
+                seeders = complete
                 leechers = incomplete
 
-                # handle the retrieved information
-                self._on_result_callback(infohash, seeders, leechers)
+                # Store the information in the dictionary
+                seed_leech_dict[infohash] = (seeders, leechers)
 
                 # remove this infohash in the infohash list of this session
                 if infohash in unprocessed_infohash_list:
                     unprocessed_infohash_list.remove(infohash)
 
         elif 'failure reason' in response_dict:
-            self._logger.debug(u"%s Failure as reported by tracker [%s]", self, repr(response_dict['failure reason']))
-
-            return False
+            self._logger.info(u"%s Failure as reported by tracker [%s]", self, repr(response_dict['failure reason']))
+            self.failed()
+            return
 
         # handle the infohashes with no result (seeders/leechers = 0/0)
         for infohash in unprocessed_infohash_list:
             seeders, leechers = 0, 0
-            # handle the retrieved information
-            self._on_result_callback(infohash, seeders, leechers)
-        return True
+            seed_leech_dict[infohash] = (seeders, leechers)
 
+        self._is_finished = True
+        self.result_deferred.callback(seed_leech_dict)
 
-class UdpTrackerSession(TrackerSession):
+    def cleanup(self):
+        """
+        Cleans the session by cancelling all deferreds and closing sockets.
+        :return: A deferred that fires once the cleanup is done.
+        """
+        super(HttpTrackerSession, self).cleanup()
+        if self.request is not None:
+            self.request.cancel()
+
+        if self.result_deferred:
+            self.result_deferred.cancel()
+
+        self.result_deferred = None
+        return defer.succeed(None)
+
+class UDPScraper(DatagramProtocol):
+
+    def __init__(self, udpsession, ip, port):
+        self._logger = logging.getLogger(self.__class__.__name__)
+        self.udpsession = udpsession
+        self.ip = ip
+        self.port = port
+        self.expect_connection_response = True
+        # Timeout after 15 seconds if nothing received.
+        self.timeout_seconds = 15
+        # The clock is used in the tests to advance time. (We are time travellers!)
+        self.clock = Clock()
```

Comment:
Hmmmm I actually did look at that one but I believe I had an issue with that approach. Let me check again.